GitCommands/Patch/Patch.cs

```diff
@@ -1,54 +1,63 @@
-using System.Text;
+﻿using System;
+using JetBrains.Annotations;
 
 namespace PatchApply
 {
-    public class Patch
+    public sealed class Patch
     {
-        private StringBuilder _textBuilder;
+        [NotNull]
+        public string Header { get; }
 
-        public Patch()
-        {
-            File = FileType.Text;
-        }
+        [CanBeNull]
+        public string Index { get; }
 
-        public enum PatchType
-        {
-            NewFile,
-            DeleteFile,
-            ChangeFile,
-            ChangeFileMode
-        }
+        public PatchFileType FileType { get; }
 
-        public enum FileType
-        {
-            Binary,
-            Text
-        }
+        [NotNull]
+        public string FileNameA { get; }
 
-        public string PatchHeader { get; set; }
-        public string PatchIndex { get; set; }
-        public FileType File { get; set; }
-        public string FileNameA { get; set; }
-        public string FileNameB { get; set; }
-        public bool CombinedDiff { get; set; }
+        [CanBeNull]
+        public string FileNameB { get; }
 
-        public PatchType Type { get; set; }
+        public bool IsCombinedDiff { get; }
 
-        public string Text => _textBuilder?.ToString();
+        public PatchChangeType ChangeType { get; }
 
-        public void AppendText(string text)
-        {
-            GetTextBuilder().Append(text);
-        }
+        [CanBeNull]
+        public string Text { get; }
 
-        public void AppendTextLine(string line)
+        public Patch(
+            [NotNull] string header,
+            [CanBeNull] string index,
+            PatchFileType fileType,
+            [NotNull] string fileNameA,
+            [CanBeNull] string fileNameB,
+            bool isCombinedDiff,
+            PatchChangeType changeType,
+            [CanBeNull] string text)
         {
-            GetTextBuilder().Append(line).Append('\n');
+            Header = header ?? throw new ArgumentNullException(nameof(header));
+            Index = index;
+            FileType = fileType;
+            FileNameA = fileNameA ?? throw new ArgumentNullException(nameof(fileNameA));
+            FileNameB = fileNameB;
+            IsCombinedDiff = isCombinedDiff;
+            ChangeType = changeType;
+            Text = text;
         }
+    }
 
-        private StringBuilder GetTextBuilder()
-        {
-            return _textBuilder ?? (_textBuilder = new StringBuilder());
-        }
+    public enum PatchChangeType
+    {
+        NewFile,
```

Comment:
`NewFile = 0,` as per https://docs.microsoft.com/en-us/dotnet/standard/design-guidelines/enum

Reply:
I'm not sure I interpret the rules at that link in the same way. From that link:

>  ✓ DO provide a value of zero on simple enums.
>
> Consider calling the value something like "None." If such a value is not appropriate for this particular enum, the most common default value for the enum should be assigned the underlying value of zero. 

This seems to suggest adding a new `None` member  and assigning it as zero, so `NewFile` then  becomes one.

I've seen this rule before and am indifferent to it. I don't understand why you'd want to add a symbol to your code that's never used and can only serve to confuse users of the API some time down the line. The literal `0` can implicitly convert to any enum type anyway. Only possible difference I see is `ToString` behaviour for a zero value if all other members are non-zero. But that hardly tips the scales to my mind.

If this is something you feel strongly about then we should add it to all enums in one go. There are plenty of violations. There's a SCA rule that requires it too, IIRC. Let me know and I can create an new issue to track it.

(In this instance I just moved the enum as-is from another place in code.)

Reply:
To me this is all about clear expression of intent.
I've come across developers who had little idea about mechanics of enums...